### PR TITLE
IDE-879 Add Tooltip for functions from metadata

### PR DESCRIPTION
**Pull request by dehilsterlexis**
Added #define NSOAP_DEBUG for future use to debug soap messaging.

### CODE REVIEW

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: comms/EclMeta.cpp [hunk 1]
```diff
@@ -231,6 +259,68 @@ bool CEclMeta::LoadImports(const std::_tstring & path, const WPathVector & folde
...
 
+bool CEclMeta::LoadFunctions(const std::_tstring & path)
+{
+    bool found = false;
+
+    for (std::map<std::wstring, StlLinked<CEclMetaData> >::iterator itr = m_masterMeta.begin(); itr != m_masterMeta.end(); ++itr)
+    {
+        if (CEclFile *file = dynamic_cast<CEclFile*>(itr->second.get()))
+        {
+            found = file->GetFunctions(m_funcs) ? true : found;
```
Comment (GordonSmith) [inline comms/EclMeta.cpp]: Should `break` at this point?

File: comms/EclMeta.cpp [hunk 2]
```diff
@@ -231,6 +259,68 @@ bool CEclMeta::LoadImports(const std::_tstring & path, const WPathVector & folde
...
+
+    CEclDefinition *func = NULL;
+    for (EclDefinitionMap::iterator itr = m_funcs.begin(); itr != m_funcs.end(); ++itr)
+    {
+        if (boost::algorithm::iequals(itr->second.get()->GetFullname(), key.GetString()) || boost::algorithm::iends_with(itr->second.get()->GetName(), key.GetString()))
+            {
+			if (itr->second.get()->GetContent().length())
+			{
+				tooltip += _T("\n");
+			}
```
Comment (GordonSmith) [inline comms/EclMeta.cpp]: Indentation

File: comms/EclMeta.h [hunk 3]
```diff
@@ -49,33 +49,91 @@ class CEclFolder : public CEclMetaData
...
+protected:
+    std::_tstring m_name;
+    std::_tstring m_type;
+public:
+    BEGIN_CUNKNOWN
+    END_CUNKNOWN(CUnknown)
+
+    CEclParam() { }
+    std::_tstring GetName() { return m_name; }
+    std::_tstring GetType() { return m_type; }
```
Comment (GordonSmith) [inline comms/EclMeta.h]: Return type should be `const TCHAR *`

File: comms/EclMeta.h [hunk 4]
```diff
@@ -49,33 +49,91 @@ class CEclFolder : public CEclMetaData
...
+    void UpdateAttrs(Element &e)
```
Comment (GordonSmith) [inline comms/EclMeta.h]: `const Element &`

File: comms/EclMeta.h [hunk 5]
```diff
@@ -49,33 +49,91 @@ class CEclFolder : public CEclMetaData
...
+    {
+        GetValue(e, _T("name"), m_name);
+        GetValue(e, _T("type"), m_type);
+    };
+    void AddType(Element e)
```
Comment (GordonSmith) [inline comms/EclMeta.h]: `const Element &`?

File: comms/EclMeta.h [hunk 6]
```diff
@@ -49,33 +49,91 @@ class CEclFolder : public CEclMetaData
...
+    {
+        GetValue(e, _T("type"), m_type);
+    }
+    void AddName(Element e)
```
Comment (GordonSmith) [inline comms/EclMeta.h]: `const Element &`?

File: comms/EclMeta.h [hunk 7]
```diff
@@ -49,33 +49,91 @@ class CEclFolder : public CEclMetaData
...
 
     CEclDefinition() { m_body = m_end = m_line = m_start = 0; }
-    CEclDefinition(const CEclDefinition & def) { m_name = def.m_name; m_type = def.m_type; m_body = def.m_body; m_end = def.m_end;  m_line = def.m_line; m_start = def.m_start; m_fields = def.m_fields; m_defs = def.m_defs; }
+    CEclDefinition(const CEclDefinition & def) { m_fullname = def.m_fullname; m_name = def.m_name; m_type = def.m_type; m_inheritType = def.m_inheritType; m_body = def.m_body; m_end = def.m_end; m_line = def.m_line; m_start = def.m_start; m_fields = def.m_fields; m_defs = def.m_defs; m_params = def.m_params; m_content = def.m_content; m_functionType = def.m_functionType; }
     std::_tstring GetName() { return m_name; }
+    std::_tstring GetFullname() { return m_fullname; }
     std::_tstring GetType() { return m_type; }
+    std::_tstring GetFunctionType() { return m_functionType; }
+    std::_tstring GetInheritType() { return m_inheritType; }
+    std::_tstring GetContent() { return m_content; }
```
Comment (GordonSmith) [inline comms/EclMeta.h]: Return types should probably be `const TCHAR *` or `const std::_tstring &`

File: comms/EclMeta.h [hunk 8]
```diff
@@ -49,33 +49,91 @@ class CEclFolder : public CEclMetaData
...
         GetIntValue(e, _T("body"), m_body);
         GetIntValue(e, _T("end"), m_end);
         GetIntValue(e, _T("line"), m_line);
         GetIntValue(e, _T("start"), m_start);
     }
+    bool HasDefinitions()
+    {
+        return m_defs.size() ? true : false;
+    }
+    void AddFunctionType(Element e)
```
Comment (GordonSmith) [inline comms/EclMeta.h]: `const Element & e`?

File: comms/EclMeta.h [hunk 9]
```diff
@@ -49,33 +49,91 @@ class CEclFolder : public CEclMetaData
...
+    {
+        GetValue(e, _T("type"), m_functionType);
+    }
+    void AddDocumentation(Element e)
```
Comment (GordonSmith) [inline comms/EclMeta.h]: `const Element & e`?

File: comms/EclMeta.h [hunk 10]
```diff
@@ -264,13 +356,16 @@ class CEclMeta
...
+    std::map<std::wstring, StlLinked<CEclDefinition> > m_funcs;
 
     bool CEclMeta::MetaExists(const std::_tstring & key);
     void PopulateMeta(const boost::filesystem::wpath & fileOrDir, const std::_tstring & dottedPath = _T(""), int level = 0);
     void Update(const WPathVector & folders, const std::wstring & xml);
     void PopulateMetaUpwards(const WPathVector & folders, const std::_tstring & path);
     CEclFile * CreateMetaFile(const std::_tstring & dottedPath, const std::_tstring & path = _T(""));
     bool LoadImports(const std::_tstring & path, const WPathVector & folders);
+    bool LoadFunctions(const std::_tstring & path);
+    std::_tstring GetFunctionTooltip(CString key);
```
Comment (GordonSmith) [inline comms/EclMeta.h]: See comments re GetTooltip above

File: wlib/LangRef.cpp [hunk 11]
```diff
@@ -741,6 +741,11 @@ class CLangRef : public ILangRef, public CUnknown
 		return true;
 	}
 
+	bool AddTooltip(const std::_tstring &label, const std::_tstring &documentation)
+	{
+		//m_langNameIndex[label].m_cat = itr->GetCategoryID();
+	}
+
```
Comment (GordonSmith) [inline wlib/LangRef.cpp]: Commented code?

File: comms/EclMeta.cpp (hunk 1 above)
Comment (dehilsterlexis) [reply comms/EclMeta.cpp]: Yes. Good catch!

Comment (dehilsterlexis) [reply comms/EclMeta.cpp]: Actually, there should not be a break here. GetFunctions accumulates the functions. The found is just a general flag that something was found.

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: comms/EclCC.h [hunk 12]
```diff
@@ -36,6 +36,7 @@ __interface IEclCC : public IUnknown
     const TCHAR * GetEclFolder(int i) const;
     bool LocatePlugin(const std::string & batchFile, boost::filesystem::path & foundFolder) const;
     bool GetAutoC(IAttribute *attr, const std::_tstring & partialLabel, StdStringVector &set);
+    std::_tstring GetToolTip(CString key);
```
[the remark this answers is not in the document]
Comment (dehilsterlexis) [reply comms/EclCC.h]: Changed

File: comms/EclCC.cpp [hunk 13]
```diff
@@ -175,6 +175,11 @@ class CEclCC : public IEclCC, public clib::CLockableUnknown
         return false;
     }
 
+    const TCHAR * GetToolTip(CString key, std::_tstring & tooltip)
```
Comment (GordonSmith) [inline comms/EclCC.cpp]: change to:  `const std::_tstring & key`

Note:  In general its preferable to use "std::_tstring" in preference to CString (and limit CString usage to marshaling between BSTR and VARIANTs) - but I do appreciate there quite a few places that use CString still.

File: comms/EclMeta.cpp (hunk 1 above)
Comment (GordonSmith) [reply comms/EclMeta.cpp]: Ok.

File: comms/EclCC.cpp (hunk 13 above)
Comment (GordonSmith) [reply comms/EclCC.cpp]: @dehilsterlexis bump

File: comms/EclCC.h [hunk 14]
```diff
@@ -36,6 +36,7 @@ __interface IEclCC : public IUnknown
     const TCHAR * GetEclFolder(int i) const;
     bool LocatePlugin(const std::string & batchFile, boost::filesystem::path & foundFolder) const;
     bool GetAutoC(IAttribute *attr, const std::_tstring & partialLabel, StdStringVector &set);
+    const TCHAR * GetToolTip(CString key, std::_tstring & tooltip);
```
Comment (GordonSmith) [inline comms/EclCC.h]: change to: `const std::_tstring & key`

File: comms/EclMeta.h (hunk 9 above)
Comment (GordonSmith) [reply comms/EclMeta.h]: @dehilsterlexis bump

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

File: comms/EclMeta.h (hunk 8 above)
Comment (GordonSmith) [reply comms/EclMeta.h]: @dehilsterlexis bump

File: comms/EclMeta.h (hunk 9 above)
Comment (GordonSmith) [reply comms/EclMeta.h]: @dehilsterlexis bump II

Comment (dehilsterlexis) [reply comms/EclMeta.h]: @Gordon so sorry! I somehow lost that.  My bad.

Comment (dehilsterlexis) [reply comms/EclMeta.h]: @GordonSmith not being able to get one const Element &e to work. It is in the AddDocumentation. Not able to solve this one.
image

**Comment by dehilsterlexis** [comment]
@GordonSmith for review

**Comment by GordonSmith** [comment]
Ready to be pulled.

**Pull request merged**